### PR TITLE
[bugfix] Resolve metrics not being properly resetted on validation epoch end

### LINKED ISSUE #9683: Getting strange validation loss/metric values when multiple data-loaders are used
### Environment
- PyTorch Lightning Version: 1.4.8 (This happens in all 1.4.x)
- PyTorch Version: 1.9.1
- Python version: 3.8
- OS: Linux
- CUDA/cuDNN version: 11.4   
- GPU models and configuration: MNIST, ddp
- How you installed PyTorch: `pip` when installing pytorch-lightning
- Any other relevant information: Till 1.3.8 everything worked

## 🐛 Bug
Hi, I have been using PL 1.3.x all along, when I updated to 1.4.x (I have tried from 1.4.0 to1.4.8) I started getting weird values for validation loss/metric. Training uses 2 gpus, `ddp` and 2 dataloaders for validation.

At `validation_epoch_end` I do aggregate (average) the results of `dataloader_idx_0` and `dataloader_idx_1`, but when I check the values printed by `self.log` they don't add up

Aggregate method used,
```py
def aggregate_validation_metrics(self, val_outputs, loss_name):
    tot_loss: torch.FloatTensor = torch.tensor(0.0, device=self.device)
    # multi data loader
    if isinstance(val_outputs[0], list):
        for loss in val_outputs:
            tot_loss += sum(loss) / len(loss)
        tot_loss = tot_loss / len(val_outputs)
    # single data loader
    else:
        tot_loss += sum(val_outputs) / len(val_outputs)

    self.log(
        f"tot_{loss_name}",
        tot_loss,
        on_step=False,
        on_epoch=True,
        prog_bar=True,
        logger=True,
        sync_dist=True,
        rank_zero_only=True,
    )
```

and its results

PL1.4.8 validation results

### Expected behavior
Correct aggregated (averaged) values at `validation_epoch_end`

### To Reproduce
I have used MNIST model and have attached the code 
1. [simple_classifier.py](https://gist.github.com/raman-rajarathinam/7ae193b8dfa1ebbd504572d2bb9ecd5b)
2. [mnist_datamodule.py](https://gist.github.com/raman-rajarathinam/1f134ffc52ce8d9e2da03c3c165b01e9)

run `python simple_classifier.py`

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9683

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 1]
```diff
@@ -214,7 +214,7 @@ def _on_evaluation_end(self, *args: Any, **kwargs: Any) -> None:
             self.trainer.call_hook("on_validation_end", *args, **kwargs)
 
         # reset any `torchmetrics.Metric` and the logger connector state
-        self.trainer.logger_connector.reset_results(metrics=True)
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:217]: The `metrics=...` flag is now unused, remove it?

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 2]
```diff
@@ -602,3 +602,69 @@ def validation_step(self, batch, batch_idx):
...
+
+            tot_loss = torch.tensor(0.0)
+            for loss in outputs:
+                tot_loss += sum(loss) / len(loss)
+            tot_loss = tot_loss / len(outputs)
+            if self.current_epoch == 0:
+                assert tot_loss == (3 + 6) / 2
+            else:
+                assert tot_loss == (30 + 60) / 2
+            self.log("tot_val_loss", tot_loss, prog_bar=True, logger=True)
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:649]: This doesn't do anything

```suggestion
```

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 3]
```diff
@@ -602,3 +602,69 @@ def validation_step(self, batch, batch_idx):
...
+                assert sum(outputs[0]) / 5 == 3
+                assert sum(outputs[1]) / 5 == 6
+            else:
+                assert sum(outputs[0]) / 5 == 30
+                assert sum(outputs[1]) / 5 == 60
+
+            tot_loss = torch.tensor(0.0)
+            for loss in outputs:
+                tot_loss += sum(loss) / len(loss)
+            tot_loss = tot_loss / len(outputs)
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:644]: This seems to pass

```suggestion
            tot_loss = torch.mean(torch.tensor(outputs, dtype=torch.float))
```

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 4]
```diff
@@ -602,3 +602,69 @@ def validation_step(self, batch, batch_idx):
...
+        def training_epoch_end(self, outputs):
+            if val_check_interval == 1.0:
+                metrics = self.trainer.progress_bar_metrics
+                if self.current_epoch == 0:
+                    assert metrics["batch_idx_epoch"] == (15 / 5.0)
+                else:
+                    assert metrics["batch_idx_epoch"] == (150 / 5.0)
+
+        def validation_step(self, batch, batch_idx, dataloader_idx):
+            value = (1 + batch_idx) * (2 if dataloader_idx == 1 else 1)
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:627]:
```suggestion
            value = (1 + batch_idx) * (1 + dataloader_idx)
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:217]: Done !

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 5]
```diff
@@ -214,7 +214,7 @@ def _on_evaluation_end(self, *args: Any, **kwargs: Any) -> None:
             self.trainer.call_hook("on_validation_end", *args, **kwargs)
 
         # reset any `torchmetrics.Metric` and the logger connector state
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:216]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 2 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:649]: Removed !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:644]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 6]
```diff
@@ -602,3 +602,69 @@ def validation_step(self, batch, batch_idx):
...
+                assert tot_loss == (3 + 6) / 2
+            else:
+                assert tot_loss == (30 + 60) / 2
+            self.log("tot_val_loss", tot_loss, prog_bar=True, logger=True)
+            assert self.trainer._results["validation_step.val_loss.0"].cumulated_batch_size == 5
+            assert self.trainer._results["validation_step.val_loss.1"].cumulated_batch_size == 5
+
+        def configure_optimizers(self):
+            return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:655]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 4 above)
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:627]: Done !

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 7]
```diff
@@ -602,3 +602,69 @@ def validation_step(self, batch, batch_idx):
...
+            self.log("batch_idx", value, on_step=True, on_epoch=True, prog_bar=True)
+            return out
+
+        def training_epoch_end(self, outputs):
+            if val_check_interval == 1.0:
+                metrics = self.trainer.progress_bar_metrics
+                if self.current_epoch == 0:
+                    assert metrics["batch_idx_epoch"] == (15 / 5.0)
+                else:
+                    assert metrics["batch_idx_epoch"] == (150 / 5.0)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply tests/trainer/logging_/test_eval_loop_logging.py:624]: Good point !

**Review by carmocca — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
